Project.ipynb

```diff
@@ -22,10 +22,1419 @@
+    "        new_dists[current] = current_time\n",
+    "        \n",
+    "        #if the arrival time is improved then the previous node is changed\n",
+    "        previous[distances > new_dists] = current\n",
+    "        \n",
+    "        #setting the correct arrival time to each node\n",
+    "        distances = np.minimum(distances, new_dists)\n",
+    "        \n",
+    "        visited_nodes += 1\n",
+    "        sorting_indexes = np.argsort(distances)\n",
+    "        \n",
+    "        current = sorting_indexes[visited_nodes]\n",
+    "    \n",
+    "    return find_next(starting_node, previous, distances, earth)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class Packet():\n",
+    "    \n",
+    "    def __init__(self, index ,destination):\n",
+    "        \n",
+    "        #final destination of the packet\n",
+    "        self.destination = destination\n",
+    "        \n",
+    "        #arrival time at the next hop\n",
+    "        self.arrival_time = -1\n",
+    "        \n",
+    "        #index of the next hop\n",
+    "        self.next_hop = -1\n",
+    "        \n",
+    "        #index that identifies the packet, used for debugging purposes\n",
+    "        self.id = index\n",
+    "        \n",
+    "        self.route = []\n",
+    "        \n",
+    "        self.arrival_times = []\n",
+    "    \n",
+    "    #function that is called when str(packet) is called\n",
+    "    def __str__(self):\n",
+    "        return f\"id: {self.id}, next hop: {self.next_hop}, destination: {self.destination}\"\n",
+    "\n",
+    "class Node():\n",
+    "    \n",
+    "    def __init__(self, packets, index):\n",
+    "        \n",
+    "        #queue of the packets\n",
+    "        self.packets = packets\n",
+    "        \n",
+    "        #index of the node\n",
+    "        self.id = index\n",
+    "    \n",
+    "#returns the queues of packets at each node\n",
+    "def get_nodes(N, packets):\n",
+    "    \n",
+    "    #generate N empty queues\n",
+    "    queues = [[] for i in range(N)]\n",
+    "    \n",
+    "    #add every packet to the queue corresponding to its starting node\n",
+    "    for i in range(packets.shape[0]):\n",
+    "        queues[packets[i,0]].append(Packet(i, packets[i,1]))\n",
+    "    \n",
+    "    #generating the Node objects\n",
+    "    nodes = []\n",
+    "    for i in range(N):\n",
+    "        nodes.append(Node(queues[i], i))\n",
+    "    \n",
+    "    return nodes \n",
+    "\n",
+    "#function used to sort the nodes based on the one that has the packet that\n",
+    "#would arrive first to the next point\n",
+    "def sorting_function(node):\n",
+    "    \n",
+    "    #if no packet is queued then the node has to be placed at the end of the sorted array\n",
+    "    if len(node.packets) == 0:\n",
+    "        return np.inf\n",
+    "    else:\n",
+    "        return node.packets[0].ar
```

Comment:
* Inside a function you should not use variables defined outside the function so pass some arguments to it
* The best way to write this function is probably:
```
propagation_time = propagation(At)
def propagation(At):
     return np.min(At, axis = 2)
```
This way the code is shorter and clearer and also the propagation time of disconnected nodes is np.inf